### DISCUSSION TITLE
Why aren't arrays rented from a shared memory pool by default?

### DISCUSSION THREAD

**Discussion by AlphaDelta** [dotnet/csharplang#6177] [Q&A] [answered]
I've been perusing through the [high performance library](https://docs.microsoft.com/en-us/windows/communitytoolkit/high-performance/introduction), specifically `MemoryOwner<T>`.

In the ['When should this be used?' section of the documentation](https://docs.microsoft.com/en-us/windows/communitytoolkit/high-performance/memoryowner#when-should-this-be-used) the following is said:

> A common use case is to replace new T[] array allocations, especially when doing repeated operations that either require a temporary buffer to work on, or that produce a buffer as a result.

Which implies that use of `MemoryOwner<T>` isn't exclusive to 'repeated operations', which I interpret to mean the initial overhead of renting an array from a shared memory pool versus allocation isn't a large concern.

A bit further down they offer an example where `new byte[]` is used in a function to read bytes from a stream, and the following is said.

> If we read a large number of files, we'll end up allocating a lot of new arrays, which will put a lot of pressure over the garbage collector

The `new byte[]` is refactored to `MemoryOwner<byte>.Allocate()` as a solution to this problem.

My question is: If renting an array from a shared memory pool is more performant than allocation, and the initial overhead of renting isn't large, why does `new T[]` not rent from a shared memory pool by default?

Is it as simple as 'the word new comes with the expectation a new object is actually being created in memory', or is it more complex than I give it credit for?

**Comment by huoyaoyuan** [comment, marked as answer]
Because rented array need to be explicitly returned. Let GC to return the array can harm the performance too.

Additionally, at which size pooling is beneficial tights to the pooling implementation. As https://github.com/dotnet/runtime/issues/68800#issuecomment-1116105818 shown, pooling small arrays can be far slower than allocating.

> My question is: If renting an array from a shared memory pool is more performant than allocation, and the initial overhead of renting isn't large

This is not true. Initial overhead of renting is very large that CoreLib refuse to use any pool other than `byte`. The overhead includes a generic instantiation, and a minimal amount of arrays cached in the pool and never reclaimed.

**Reply by AlphaDelta** [reply]
Thank you for your reply and that interesting issue https://github.com/dotnet/runtime/issues/68800#issuecomment-1116105818, very informative!

**Comment by shandianmao110** [comment]
Your e-mail has been received. I will reply later.
